File: csharp/uwapi/map.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Unnatural
{
    using MapStateEnum = Interop.UwMapStateEnum;
    using StartingPosition = Interop.UwMapStartingPosition;

    public struct Vector3
    {
        public float x, y, z;
    }

    public static class Map
    {
        public static string Name()
        {
            return name;
        }

        public static string Guid()
        {
            return guid;
        }

        public static string Path()
        {
            return path;
        }

        public static uint MaxPlayers()
        {
            return maxPlayers;
        }

        public static IReadOnlyList<StartingPosition> StartingPositions()
        {
            return startingPositions;
        }

        public static IReadOnlyList<Vector3> Positions()
        {
            return positions;
        }

        public static Vector3 Position(uint position)
        {
            return positions[(int)position];
        }

        public static IReadOnlyList<Vector3> Ups()
        {
            return ups;
        }

        public static Vector3 Up(uint position)
        {
            return ups[(int)position];
        }

        public static IReadOnlyList<IReadOnlyList<uint>> Neighbors()
        {
            return neighbors;
        }

        public static IReadOnlyList<uint> Neighbors(uint position)
        {
            return neighbors[(int)position];
        }

        public static IReadOnlyList<byte> Terrains()
        {
            return terrains;
        }

        public static byte Terrain(uint position)
        {
            return terrains[(int)position];
        }

        public static uint[] AreaRange(Vector3 point, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaRange(point.x, point.y, point.z, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static uint[] AreaConnected(uint position, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaConnected(position, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static uint[] AreaNeighborhood(uint position, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaNeighborhood(position, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static uint[] AreaExtended(uint position, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaExtended(position, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static bool TestVisible(Vector3 a, Vector3 b)
        {
            return Interop.uwTestVisible(a.x, a.y, a.z, b.x, b.y, b.z);
        }

        public static bool TestShooting(uint shooterPosition, uint shooterProto, float shootingRangeUpgrade, uint targetPosition, uint targetProto)
        {
            return Interop.uwTestShooting(shooterPosition, shooterProto, shootingRangeUpgrade, targetPosition, targetProto);
        }

        public static float DistanceLine(uint a, uint b)
        {
            Vector3 a3 = positions[(int)a];
            Vector3 b3 = positions[(int)b];
            float x = a3.x - b3.x;
            float y = a3.y - b3.y;
            float z = a3.z - b3.z;
            x *= x;
            y *= y;
            z *= z;
            return (float)Math.Sqrt(x + y + z);
        }

        public static float DistanceEstimate(uint a, uint b)
        {
            return Interop.uwDistanceEstimate(a, b);
        }

        public static float Yaw(uint a, uint b)
        {
            return Interop.uwYaw(a, b);
        }

        public static IReadOnlyList<uint> TileToCluster()
        {
            return mapTileToCluster;
        }

        public static IReadOnlyList<uint> ClusterToTile()
        {
            return mapClusterToTile;
        }

        public static uint TileToCluster(uint tile)
        {
            return mapTileToCluster[(int)tile];
        }

        public static uint ClusterToTile(uint cluster)
        {
            return mapClusterToTile[(int)cluster];
        }

        public static IReadOnlyList<IReadOnlyList<uint>> ClustersNeighbors()
        {
            return clustersNeighbors;
        }

        public static IReadOnlyList<uint> ClustersNeighbors(uint cluster)
        {
            return clustersNeighbors[(int)cluster];
        }

        static string name;
        static string guid;
        static string path;
        static uint maxPlayers;
        static readonly List<StartingPosition> startingPositions = new List<StartingPosition>();
        static readonly List<Vector3> positions = new List<Vector3>();
        static readonly List<Vector3> ups = new List<Vector3>();
        static readonly List<uint[]> neighbors = new List<uint[]>();
        static readonly List<byte> terrains = new List<byte>();
        static readonly List<uint> mapTileToCluster = new List<uint>();
        static readonly List<uint> mapClusterToTile = new List<uint>();
        static readonly List<uint[]> clustersNeighbors = new List<uint[]>();

        static void Load()
        {
            Game.LogInfo("loading map");

            startingPositions.Clear();
            positions.Clear();
            ups.Clear();
            neighbors.Clear();
            terrains.Clear();
            mapTileToCluster.Clear();
            mapClusterToTile.Clear();
            clustersNeighbors.Clear();

            {
                Interop.UwMapInfo info = new Interop.UwMapInfo();
                Interop.uwMapInfo(ref info);
                name = Marshal.PtrToStringAnsi(info.name);
                guid = Marshal.PtrToStringAnsi(info.guid);
                path = Marshal.PtrToStringAnsi(info.path);
                maxPlayers = info.maxPlayers;
            }

            {
                Interop.UwMapStartingPositionsArray data = new Interop.UwMapStartingPositionsArray();
                Interop.uwMapStartingPositions(ref data);
                int size = Marshal.SizeOf(typeof(StartingPosition));
                for (int i = 0; i < data.count; i++)
                {
                    IntPtr currentPtr = IntPtr.Add(data.data, i * size);
                    var tmp = Marshal.PtrToStructure<StartingPosition>(currentPtr);
                    startingPositions.Add(tmp);
                }
            }

            {
                uint count = Interop.uwTilesCount();
                Interop.UwTile tile = new Interop.UwTile();
                for (uint i = 0; i < count; ++i)
                {
                    Interop.uwTile(i, ref tile);
                    {
                        Vector3 p = new Vector3();
                        p.x = tile.position[0];
                        p.y = tile.position[1];
                        p.z = tile.position[2];
                        positions.Add(p);
                    }
                    {
                        Vector3 u = new Vector3();
                        u.x = tile.up[0];
                        u.y = tile.up[1];
                        u.z = tile.up[2];
                        ups.Add(u);
                    }
                    {
                        uint[] tmp = new uint[tile.neighborsCount];
                        if (tile.neighborsCount > 0)
                            Marshal.Copy(tile.neighborsIndices, (int[])(object)tmp, 0, (int)tile.neighborsCount);
                        neighbors.Add(tmp);
                    }
                    terrains.Add(tile.terrain);
                    mapTileToCluster.Add(tile.clusterIndex);
                }
            }

            {
                uint count = Interop.uwClustersCount();
                Interop.UwCluster cluster = new Interop.UwCluster();
                for (uint i = 0; i < count; ++i)
                {
                    Interop.uwCluster(i, ref cluster);
                    {
                        uint[] tmp = new uint[cluster.neighborsCount];
                        if (cluster.neighborsCount > 0)
                            Marshal.Copy(cluster.neighborsIndices, (int[])(object)tmp, 0, (int)cluster.neighborsCount);
                        clustersNeighbors.Add(tmp);
                    }
                    mapClusterToTile.Add(cluster.centerTileIndex);
                }
            }

            Game.LogInfo("map loaded");
        }

        static void MapStateChanged(object sender, MapStateEnum state)
        {
            if (state == MapStateEnum.Loaded)
                Load();
        }

        static Map()
        {
            Game.MapStateChanged += MapStateChanged;
        }
    }
}

File: python/uwapi/map.py
import math
from dataclasses import dataclass
from .interop import *
from .events import uw_events

INVALID: int = 0xFFFFFFFF

@dataclass
class Vector3:
    x: float = 0
    y: float = 0
    z: float = 0

class Map:
    _instance = None
    _name: str = ""
    _guid: str = ""
    _path: str = ""
    _max_players: int = 0
    _starting_positions: List[UwMapStartingPosition] = []
    _positions: List[Vector3] = []
    _ups: List[Vector3] = []
    _neighbors: List[List[int]] = []
    _terrains: List[int] = []
    _map_tile_to_cluster: List[int] = []
    _map_cluster_to_tile: List[int] = []
    _clusters_neighbors: List[List[int]] = []

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            uw_events.on_map_state(cls._instance._map_state)
        return cls._instance

    def name(self) -> str:
        return self._name

    def guid(self) -> str:
        return self._guid

    def path(self) -> str:
        return self._path

    def max_players(self) -> int:
        return self._max_players

    def starting_positions(self) -> List[UwMapStartingPosition]:
        return self._starting_positions

    def positions(self) -> List[Vector3]:
        return self._positions

    def position(self, position: int) -> Vector3:
        return self._positions[position]

    def ups(self) -> List[Vector3]:
        return self._ups

    def up(self, position: int) -> Vector3:
        return self._ups[position]

    def neighbors_all(self) -> List[List[int]]:
        return self._neighbors

    def neighbors(self, position: int) -> List[int]:
        return self._neighbors[position]

    def terrains(self) -> List[int]:
        return self._terrains

    def terrain(self, position: int) -> int:
        return self._terrains[position]

    def area_range(self, point: Vector3, radius: float) -> List[int]:
        return uw_interop.uwAreaRange(point.x, point.y, point.z, radius).ids

    def area_connected(self, position: int, radius: float) -> List[int]:
        return uw_interop.uwAreaConnected(position, radius).ids

    def area_neighborhood(self, position: int, radius: float) -> List[int]:
        return uw_interop.uwAreaNeighborhood(position, radius).ids

    def area_extended(self, position: int, radius: float) -> List[int]:
        return uw_interop.uwAreaExtended(position, radius).ids

    def test_visible(self, a: Vector3, b: Vector3) -> bool:
        return uw_interop.uwTestVisible(a.x, a.y, a.z, b.x, b.y, b.z)

    def test_shooting(
        self,
        shooter_position: int,
        shooter_proto: int,
        shooting_range_upgrade: float,
        target_position: int,
        target_proto: int
    ) -> bool:
        return uw_interop.uwTestShooting(
            shooter_position, shooter_proto, shooting_range_upgrade,
            target_position, target_proto
        )

    def distance_line(self, a: int, b: int) -> float:
        a3 = self._positions[a]
        b3 = self._positions[b]
        dx = a3.x - b3.x
        dy = a3.y - b3.y
        dz = a3.z - b3.z
        return math.sqrt(dx * dx + dy * dy + dz * dz)

    def distance_estimate(self, position_a: int, position_b: int) -> float:
        return uw_interop.uwDistanceEstimate(position_a, position_b)

    def yaw(self, start_position: int, goal_position: int) -> float:
        return uw_interop.uwYaw(start_position, goal_position)

    def tile_to_cluster_map(self) -> List[int]:
        return self._map_tile_to_cluster

    def cluster_to_tile_map(self) -> List[int]:
        return self._map_cluster_to_tile

    def tile_to_cluster(self, tile: int) -> int:
        return self._map_tile_to_cluster[tile]

    def cluster_to_tile(self, cluster: int) -> int:
        return self._map_cluster_to_tile[cluster]

    def clusters_neighbors_all(self) -> List[List[int]]:
        return self._clusters_neighbors

    def clusters_neighbors(self, cluster: int) -> List[int]:
        return self._clusters_neighbors[cluster]

    def _reset(self) -> None:
        self._name: str = ""
        self._guid: str = ""
        self._path: str = ""
        self._max_players: int = 0
        self._starting_positions: List[UwMapStartingPosition] = []
        self._positions: List[Vector3] = []
        self._ups: List[Vector3] = []
        self._neighbors: List[List[int]] = []
        self._terrains: List[int] = []
        self._map_tile_to_cluster: List[int] = []
        self._map_cluster_to_tile: List[int] = []
        self._clusters_neighbors: List[List[int]] = []

    def _load_info(self) -> None:
        info = uw_interop.uwMapInfo()
        if info[0]:
            self._name = info[1].name;
            self._guid = info[1].guid;
            self._path = info[1].path;
            self._max_players = info[1].maxPlayers

    def _load_tiles(self) -> None:
        count = uw_interop.uwTilesCount()
        for i in range(count):
            tile = uw_interop.uwTile(i)
            self._positions.append(Vector3(tile.position[0], tile.position[1], tile.position[2]))
            self._ups.append(Vector3(tile.up[0], tile.up[1], tile.up[2]))
            self._neighbors.append(tile.neighborsIndices)
            self._terrains.append(tile.terrain)
            self._map_tile_to_cluster.append(tile.clusterIndex)

    def _load_clusters(self) -> None:
        count = uw_interop.uwClustersCount()
        for i in range(count):
            cluster = uw_interop.uwCluster(i)
            self._clusters_neighbors.append(cluster.neighborsIndices)
            self._map_cluster_to_tile.append(cluster.centerTileIndex)

    def _load(self) -> None:
        uw_interop.uwLog(UwSeverityEnum.Info, "loading map")
        self._reset()
        self._load_info()
        self._starting_positions = uw_interop.uwMapStartingPositions().data
        self._load_tiles()
        self._load_clusters()
        uw_interop.uwLog(UwSeverityEnum.Info, "map loaded")

    def _map_state(self, state: UwMapStateEnum) -> None:
        if state == UwMapStateEnum.Loaded:
            self._load()

uw_map = Map()
